### PR TITLE
Support prescribed labels 

### PR DESCRIPTION
**Pull request by Konstantinov-Innokentii**
# What this PR does

**Cleanup label typing:**
1. LabelParam -> two separate types LabekKey and LabelValue 
2. LabelData -> renamed to LabelPair. 
3. LabelKeyData -> renamed to LabelOption
Data is not giving any info about what this type represents. 
4. Remove LabelsData and LabelsKeysData types. They are just list of types listed above and with new naming it feels obsolete.
5. ValueData removed. LabelPair is used instead.
6. Rework  AlertGroupCustomLabel to use LabelKey type for key to make type system more consistent. Name model type  AlertGroupCustomLabel**DB** and api type  AlertGroupCustomLabel**API** to clearly distinguish them.

**Split update_labels_cache into two tasks** update_label_option_cache and update_label_pairs_cache.
Original task was expecting array of  LabelsData (now it's LabelPair) OR one LabelKeyData ( now it's LabelOption). I believe having one function with two sp different argument types makes it more complicated for understanding. 

**Make OnCall backend support prescribed labels**. OnCall will sync and store "prescribed" field for key and values, so Label dropdown able to disable editing for certain labels.

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [ ] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: engine/apps/labels/utils.py [hunk 1]
```diff
@@ -16,32 +16,28 @@
...
 
-class LabelUpdateParam(typing.TypedDict):
+class Key(typing.TypedDict):
+    id: str
     name: str
+    prescribed: bool
 
 
-class LabelParams(typing.TypedDict):
+class Value(typing.TypedDict):
```
Comment (Ferril) [inline engine/apps/labels/utils.py:25]: I'd propose to rename it to `LabelValue`

File: engine/apps/labels/tasks.py [hunk 2]
```diff
@@ -64,6 +72,61 @@ def update_labels_cache(labels_data: LabelsData | LabelKeyData):
...
+
+        if value.key.name != values_id_to_pair[value.id]["key"]["name"]:
+            value.key.name = values_id_to_pair[value.id]["key"]["name"]
+        if value.key.prescribed != values_id_to_pair[value.id]["key"]["prescribed"]:
+            value.key.prescribed = values_id_to_pair[value.id]["key"]["prescribed"]
+        value.key.last_synced = now
+        keys_to_update.add(value.key)
+
+    LabelKeyCache.objects.bulk_update(keys_to_update, fields=["name", "last_synced", "prescribed"])
+    LabelKeyCache.objects.bulk_update(values, fields=["name", "last_synced", "prescribed"])
```
Comment (Ferril) [inline engine/apps/labels/tasks.py:127]: typo: the second update should be `LabelValueCache`

File: engine/apps/labels/models.py [hunk 3]
```diff
@@ -71,16 +75,19 @@ def update_association(labels_data: "LabelsData", instance: models.Model, organi
...
-            value_name = label_data["value"]["name"]
+        for label_pair in label_pairs:
+            key_id = label_pair["key"]["id"]
+            key_name = label_pair["key"]["name"]
+            key_prescribed = label_pair["key"]["prescribed"]
 
-            label_key = LabelKeyCache(id=key_id, name=key_name, organization=organization)
+            value_id = label_pair["value"]["id"]
+            value_name = label_pair["value"]["name"]
+            value_prescribed = label_pair["key"]["prescribed"]
```
Comment (Ferril) [inline engine/apps/labels/models.py:85]:
```suggestion
            value_prescribed = label_pair["value"]["prescribed"]
```

File: engine/apps/api/views/labels.py [hunk 4]
```diff
@@ -51,88 +51,95 @@ class LabelsViewSet(LabelsFeatureFlagViewSet):
...
     @extend_schema(responses=LabelValueSerializer)
     def get_value(self, request, key_id, value_id):
-        """Value name"""
+        """get_value returns a Value"""
         organization = self.request.auth.organization
-        result, response = LabelsAPIClient(organization.grafana_url, organization.api_token).get_value(key_id, value_id)
-        self._update_labels_cache(result)
-        return Response(result, status=response.status_code)
+        value, response = LabelsAPIClient(organization.grafana_url, organization.api_token).get_value(key_id, value_id)
+        # TODO: update_labels_cache expects LabelOption, but get value returns a Value. Investigate, temporary disable.
```
Comment (vadimkerr) [inline engine/apps/api/views/labels.py:74]: Does this TODO need to be addressed before merging the PR?

File: engine/apps/labels/models.py [hunk 5]
```diff
@@ -89,12 +97,15 @@ def update_association(labels_data: "LabelsData", instance: models.Model, organi
...
+        # create labels cache and associations that don't exist.
+        # Ignoring conflicts because some labels might laready exist. They will be updates in task.
         LabelKeyCache.objects.bulk_create(labels_keys, ignore_conflicts=True, batch_size=5000)
         LabelValueCache.objects.bulk_create(labels_values, ignore_conflicts=True, batch_size=5000)
         instance.labels.model.objects.bulk_create(labels_associations, ignore_conflicts=True, batch_size=5000)
 
-        update_labels_cache.apply_async((labels_data,))
+        # Many labels
+        update_label_pairs_cache.apply_async((label_pairs,))
+        # update_labels_cache.apply_async((label_pairs,))
```
Comment (vadimkerr) [inline engine/apps/labels/models.py:108]: Looks like this line should be deleted

**Review by vadimkerr — APPROVED** [review]
LGTM. Consider adding simple unit tests for the new Celery tasks

File: engine/apps/api/views/labels.py (hunk 4 above)
Comment (Konstantinov-Innokentii) [reply engine/apps/api/views/labels.py:74]: No, it's for future iterations.

File: engine/apps/labels/models.py (hunk 5 above)
Comment (Konstantinov-Innokentii) [reply engine/apps/labels/models.py:108]: thanks

**Comment by joeyorlando** [comment]
just FYI for next time, this PR introduced linting issues into `dev`

**Comment by Konstantinov-Innokentii** [comment]
@joeyorlando Oh, I thought it was introduced [here](https://github.com/grafana/oncall/actions/runs/7963261048/job/21738463517). Did I bring new one?

**Pull request merged**